File: backend/internal/domain/component/component.go
package component

import "time"

// Role categorises a component for meal planning.
type Role string

const (
	RoleMain        Role = "main"
	RoleSideStarch  Role = "side_starch"
	RoleSideVeg     Role = "side_veg"
	RoleSideProtein Role = "side_protein"
	RoleSauce       Role = "sauce"
	RoleDrink       Role = "drink"
	RoleDessert     Role = "dessert"
	RoleStandalone  Role = "standalone"
)

var validRoles = map[Role]bool{
	RoleMain:        true,
	RoleSideStarch:  true,
	RoleSideVeg:     true,
	RoleSideProtein: true,
	RoleSauce:       true,
	RoleDrink:       true,
	RoleDessert:     true,
	RoleStandalone:  true,
}

// ValidRole reports whether r is a known component role.
func ValidRole(r Role) bool { return validRoles[r] }

// Component is the aggregate root for a dish/recipe.
type Component struct {
	ID                int64
	Name              string
	Role              Role
	VariantGroupID    *int64
	ReferencePortions float64
	PrepMinutes       *int
	CookMinutes       *int
	ImagePath         *string
	Notes             *string
	LastCookedAt      *time.Time
	CookCount         int
	Favorite          bool
	Ingredients       []ComponentIngredient
	Instructions      []Instruction
	Tags              []string
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

// ComponentIngredient links an ingredient to a component with a quantity.
// IngredientName is populated when loaded through the repo; it is not persisted
// back on write.
//
// GramsSource explains how Grams was resolved for the current in-memory value.
// It is set by the resolver at write time and is not persisted.
type ComponentIngredient struct {
	ID             int64
	ComponentID    int64
	IngredientID   int64
	IngredientName string
	Amount         float64
	Unit           string
	Grams          float64
	GramsSource    string
	SortOrder      int
}

// Grams-resolution sources surfaced by the service to help clients show
// confidence badges. These are not persisted to the database.
const (
	GramsSourceDirect   = "direct"   // unit is a mass (g/kg/mg/oz/lb), exact
	GramsSourcePortion  = "portion"  // matched an ingredient-specific portion
	GramsSourceDefault  = "default"  // universal mass default (e.g., "oz")
	GramsSourceFallback = "fallback" // universal volume default (water-density)
	GramsSourceManual   = "manual"   // user-supplied grams; unit could not be auto-resolved
)

// Instruction is a numbered cooking step.
type Instruction struct {
	ID          int64
	ComponentID int64
	StepNumber  int
	Text        string
}

// ListQuery holds filtering, pagination, and sorting for component listing.
type ListQuery struct {
	Search       string
	Role         string
	Tag          string
	FavoriteOnly bool
	Limit        int
	Offset       int
	SortBy       string
	SortDesc     bool
}

// ListResult wraps a page of components with the total count.
type ListResult struct {
	Items []Component
	Total int
}

// InsightsQuery parameterises the rotation-insights lookup. Zero values are
// replaced with defaults by Service.Insights.
type InsightsQuery struct {
	ForgottenWeeks  int
	ForgottenLimit  int
	MostCookedLimit int
}

// Insights bundles the two rotation signals used by the component library
// badges and archive view.
type Insights struct {
	Forgotten  []Component
	MostCooked []Component
}

File: frontend/src/lib/schemas/component.ts
import { z } from "zod/v4"

export const COMPONENT_ROLES = [
  "main",
  "side_starch",
  "side_veg",
  "side_protein",
  "sauce",
  "drink",
  "dessert",
  "standalone",
] as const

export const componentIngredientSchema = z.object({
  ingredient_id: z.coerce.number().positive(),
  ingredient_name: z.string().optional(),
  amount: z.coerce.number().positive(),
  unit: z.string().min(1),
  grams: z.coerce.number().min(0).default(0),
  sort_order: z.coerce.number().min(0).default(0),
  // Per-100g macros stored client-side for live nutrition preview
  kcal_100g: z.coerce.number().min(0).default(0).optional(),
  protein_100g: z.coerce.number().min(0).default(0).optional(),
  fat_100g: z.coerce.number().min(0).default(0).optional(),
  carbs_100g: z.coerce.number().min(0).default(0).optional(),
  fiber_100g: z.coerce.number().min(0).default(0).optional(),
  sodium_100g: z.coerce.number().min(0).default(0).optional(),
})

export const instructionSchema = z.object({
  step_number: z.coerce.number().positive(),
  text: z.string().min(1),
})

export const componentSchema = z.object({
  name: z.string().min(1),
  role: z.enum(COMPONENT_ROLES),
  reference_portions: z.coerce.number().positive().default(1),
  prep_minutes: z.coerce.number().min(0).default(0),
  cook_minutes: z.coerce.number().min(0).default(0),
  notes: z.string().nullable().optional(),
  ingredients: z.array(componentIngredientSchema).min(1).default([]),
  instructions: z.array(instructionSchema).min(1).default([]),
  tags: z.array(z.string().min(1)).default([]),
})

export type ComponentFormValues = z.infer<typeof componentSchema>
